snapcraft/cli/_channel_map.py

```diff
@@ -0,0 +1,267 @@
+    )
+
+    try:
+        progressive_mapped_channel: Optional[
+            MappedChannel
+        ] = snap_channel_map.get_mapped_channel(
+            channel_name=channel_name, architecture=architecture, progressive=True
+        )
+    except ValueError:
+        progressive_mapped_channel = None
+
+    if progressive_mapped_channel is not None:
+        progressive_revision = snap_channel_map.get_revision(
+            progressive_mapped_channel.revision
+        )
+        progressive_mapped_channel_line = _get_channel_line(
+            mapped_channel=progressive_mapped_channel,
+            revision=progressive_revision,
+            channel_info=channel_info,
+            hint=hint,
+            progress_string=f"{_HINTS.PROGRESSING_TO} {progressive_mapped_channel.progressive.percentage:.0f}%",
+        )
+        # Setup progress for the actually released revision, this needs to be
+        # calculated. But only show it if the channel is open.
+        progress_string = "{} {:.0f}%".format(
+            _HINTS.PROGRESSING_TO,
+            100 - progressive_mapped_channel.progressive.percentage,
+        )
+    else:
+        progress_string = _HINTS.NO_PROGRESS
+
+    try:
+        mapped_channel: Optional[MappedChannel] = snap_channel_map.get_mapped_channel(
+            channel_name=channel_name, architecture=architecture, progressive=False
+        )
+    except ValueError:
+        mapped_channel = None
+
+    if mapped_channel is not None:
+        revision = snap_channel_map.get_revision(mapped_channel.revision)
+        channel_lines.append(
+            _get_channel_line(
+                mapped_channel=mapped_channel,
+                revision=revision,
+                channel_info=channel_info,
+                hint=hint,
+                progress_string=progress_string,
+            )
+        )
+    # Don't show empty branches
+    elif channel_info.branch is None:
+        channel_lines.append(
+            _get_channel_line(
+                mapped_channel=None,
+                revision=None,
+                channel_info=channel_info,
+                hint=hint,
+                progress_string=_HINTS.NO_PROGRESS
+                if hint == _HINTS.CLOSED
+                else progress_string,
+            )
+        )
+
+    if (
+        os.getenv("SNAPCRAFT_EXPERIMENTAL_PROGRESSIVE_DELIVERY")
+        and progressive_mapped_channel is not None
+    ):
+        channel_lines.append(progressive_mapped_channel_line)
+
+    return channel_lines
+
+
+def _has_channels_for_architecture(
+    snap_channel_map, architecture: str, channels: List[str]
+) -> bool:
+    for channel_name in channels:
+        for progressive in (False, True):
+            try:
+                snap_channel_map.get_mapped_channel(
+                    channel_name=channel_name,
+                    architecture=architecture,
+                    progressive=progressive,
+                )
+                return True
```

Comment:
uh oh, here's a case of multiple returns ;)

Reply:
fixed